tools/doc/apilinks.js

```diff
@@ -0,0 +1,163 @@
+'use strict';
+
+// Scan API sources for definitions.
+//
+// Note the output is produced based on a world class parser, adherence to
+// conventions, and a bit of guess work.  Examples:
+//
+//  * We scan for top level module.exports statements, and determine what
+//    is exported by looking at the source code only (i.e., we don't do
+//    an eval).  If exports include `Foo`, it probably is a class, whereas
+//    if what is exported is `constants` it probably is prefixed by the
+//    by the basename of the source file (e.g., `zlib`), unless that
+//    source file is `buffer.js`, in which case the name is just `buf`.
+//    unless the constant is `kMaxLength`, in which case it is `buffer`.
+//
+//  * We scan for top level definitions for those exports, handling
+//    most common cases (e.g., `X.prototype.foo =`, `X.foo =`,
+//    `function X(...) {...}`).  Over time, we expect to handle more
+//    cases (example: ES2015 class definitions).
+
+const acorn = require('acorn');
+const fs = require('fs');
+const path = require('path');
+const child_process = require('child_process');
+
+// determine orign repo and most recent commit
+const repo = (
+  child_process.execSync(
+    'git config remote.origin.url',
+    { stdio: ['ignore', null, 'ignore'] }
+  ).toString()
+ .match(/(\w+\/\w+)\.git\r?\n?$/) || ['', 'nodejs/node'])[1];
+
+let hash = child_process.execSync(
+  'git log -1 --pretty=%H',
+  { stdio: ['ignore', null, 'ignore'] }
+).toString().trim() || 'master';
+
+try {
+  // replace hash with tag name, if tagged
+  hash = child_process.execSync(
+    `git describe --contains ${hash}`,
+    { stdio: ['ignore', null, 'ignore'] }
+  ).toString();
+} catch {
+}
+
+const definition = {};
+process.argv.slice(2).forEach((file) => {
+  const basename = path.basename(file, '.js');
+
+  // parse source
+  const source = fs.readFileSync(file, 'utf8');
+  const ast = acorn.parse(source, { ecmaVersion: 10 });
```

Comment:
adding `locations: true` will annotate the nodes with line & column location. So no need for `lineends`.
e.g:
```js
> const acorn = require('internal/deps/acorn/dist/acorn');
> const source = fs.readFileSync(file, 'utf8');
> const ast = acorn.parse('lib/buffer.js', { ecmaVersion: 10, locations: true })
> console.log(ast.body[2].loc)
SourceLocation {
  start: Position { line: 42, column: 0 },
  end: Position { line: 42, column: 21 } }
```

Reply:
or `const acorn = require('../../deps/acorn');`

Reply:
That gets me:

```
SyntaxError: Unexpected token (127:12)
    at Parser.pp$4.raise (/Users/rubys/git/node/deps/acorn/dist/acorn.js:2649:13)
    at Parser.pp.unexpected (/Users/rubys/git/node/deps/acorn/dist/acorn.js:637:8)
    at Parser.pp.expect (/Users/rubys/git/node/deps/acorn/dist/acorn.js:631:26)
    at Parser.pp$1.parseTryStatement (/Users/rubys/git/node/deps/acorn/dist/acorn.js:983:10)
    at Parser.pp$1.parseStatement (/Users/rubys/git/node/deps/acorn/dist/acorn.js:768:32)
    at Parser.pp$1.parseBlock (/Users/rubys/git/node/deps/acorn/dist/acorn.js:1077:23)
    at Parser.pp$1.parseStatement (/Users/rubys/git/node/deps/acorn/dist/acorn.js:775:34)
    at Parser.pp$1.parseIfStatement (/Users/rubys/git/node/deps/acorn/dist/acorn.js:902:26)
    at Parser.pp$1.parseStatement (/Users/rubys/git/node/deps/acorn/dist/acorn.js:764:31)
    at Parser.pp$1.parseBlock (/Users/rubys/git/node/deps/acorn/dist/acorn.js:1077:23)
make[1]: *** [out/apilinks.json] Error 1
```

This is while parsing [`lib/assert.js`](https://github.com/nodejs/node/blob/c8a27a761773b9815bd6af07ce6d6e72ae19df30/lib/assert.js#L127)

dep/acorn is at version [5.2.1](https://github.com/nodejs/node/blob/c8a27a761773b9815bd6af07ce6d6e72ae19df30/deps/acorn/package.json#L7), and needs to be at [5.6.0](https://github.com/acornjs/acorn/blob/master/CHANGELOG.md#560-2018-05-31) or later to support this relatively new JavaScript feature.

How did you get past this?